### DISCUSSION TITLE
Clarify how MAUI will use .NET Dependency Injection

### DISCUSSION THREAD

**Discussion by matt-goldman** [dotnet/maui#1237] [Q&A] [answered]
I'm playing with the new built in DI capability, so in Startup I've registered some services:

```
public void Configure(IAppHostBuilder appBuilder)
{
	appBuilder
	.UseFormsCompatibility()
	.UseMauiApp<App>()
	.ConfigureServices(services => 
	{
		services.AddTransient<IMyService, MyService>();
                services.AddSingleton<MyOtherService>();
	})
	.ConfigureFonts(fonts =>
	{
		fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
	});
}
```

How do I then consume these?

One way might be to use ctor injection to get IServiceProvider into my App.xaml.cs. I could then just either instantiate my services in here or insatiate a static instance of the service provider, which I can then consume from elsewhere in the app to resolve dependencies.

```
public static IServiceProvider serviceProvider { get; set; }
		
public App(IServiceProvider sp)
{
	InitializeComponent();
	serviceProvider = sp;
}
```

And then I can use `App.serviceProvider.GetService...` to resolve them in my viewmodels (or use ctor injection in the viewmodels and use it to resolve the viewmodel). Although I would probably have a static IServiceScope rather than service provider. Either way, this seems a bit hacky.

An alternative to the above suggestion is to have a static/singleton class in the app purely for dependency resolution. This is the approach used in the eShopOnContainers sample app, with TinyIoC in the [ViewModelLocator service](https://github.com/dotnet-architecture/eshop-mobile-client/blob/main/eShopOnContainers/eShopOnContainers.Core/ViewModels/Base/ViewModelLocator.cs).

Is this the right way? Is there a recommended or supported way to do this in MAUI?

**Comment by matt-goldman** [comment]
Further to the above question - using a 3rd party IoC container or MVVM library, I can resolve a *ViewModel* including it's dependencies. For example, let's say I have the following constructor:

```
public MyViewModel(IMyService myService, IMyOtherService, otherService)
{
     ...
}
```

Using something like TinyIoC or Autofac, I can call something like `container.Resolve<MyViewModel>();` which will return a fully constructed instance of MyViewModel with all it's required dependencies. Using my example above, even with access to the Service Provider, I can use it to resolve the services but still have to instantiate my viewmodel manually:

```
var myService = App.scope.ServiceProvicer.GetService<IMyService>();
var otherService = App.scope.ServiceProvicer.GetService<IMyOtherService>();

var vm = new MyViewModel(myService, otherService);
```

Is this the point where you say maybe a third party container is the way to go? And if that is the case, what is the use case for including the built-in DI? Will it just be to replace the DependencyService for registering platform specific implementations?

**Reply by bakerhillpins** [reply]
> Using something like TinyIoC or Autofac, I can call something like `container.Resolve<MyViewModel>();` which will return a fully constructed instance of MyViewModel with all it's required dependencies. Using my example above, even with access to the Service Provider, I can use it to resolve the services but still have to instantiate my viewmodel manually:

@matt-goldman  At the risk of duplication, I'm [linking to a comment](https://github.com/dotnet/maui/issues/3622#issuecomment-1144042313) I made regarding exactly this situation and how it's reduced my VM code and complexity.

I'd be happy to discuss this further or even provide a contrived example. I'm also looking at your plugin to see if this can be included, at least temporarily.  Otherwise, I'll need a 3rd party solution or end up rewriting a bunch of VMs.

**Comment by matt-goldman** [comment, marked as answer]
Not sure if this will be useful to anyone else - but I put this together and its working for me: https://github.com/matt-goldman/Maui.Plugins.PageResolver

**Comment by pjmagee** [comment]
Probably should be a new discussion but somewhat related to the DI. I'm having issues running a  ` services.AddHostedService<Service>();` from `Microsoft.Extensions.Hosting`. I thought Maui would hook into the normal  GenericHost` but I'm having issues running a background service. Is it something to do with the way the UI is initialized.

**Comment by pjmagee** [comment]
I came across an issue where I was using a Blazor Library that was made for Web, but it registered services as `Scoped` which meant the components for me failed to work. I wonder if there is any way to override `Scoped` registration to become `Singleton` for situations like this. Or if we just will need to manually remove and re-register services from Libraries we use.

**Reply by fairking** [reply]
Do you have an access to `ServiceCollection`? After all your services have been registered you can easily take them out of the `ServiceCollection` and register them again as Singletons.

``` c#
public static class MauiProgram
{
	public static MauiApp CreateMauiApp()
	{
		var builder = MauiApp.CreateBuilder();
		builder
			.UseMauiApp<App>()
			.ConfigureFonts(fonts =>
			{
				fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
				fonts.AddFont("OpenSans-Semibold.ttf", "OpenSansSemibold");
			});

		builder.Services.SetupServicesByExternalLibrary(); // Your external library you have no control of which produces scoped services

                // Change scoped lifetime to singleton
		var scopedServices = builder.Services.Where(x => x.Lifetime == ServiceLifetime.Scoped).ToList();
		foreach (var s in scopedServices)
                {
			builder.Services.Remove(s);

			if (s.ImplementationFactory != null)
				builder.Services.Add(new ServiceDescriptor(s.ServiceType, s.ImplementationFactory, ServiceLifetime.Singleton));
			else if (s.ImplementationInstance != null)
				builder.Services.Add(new ServiceDescriptor(s.ServiceType, s.ImplementationInstance));
			else
				builder.Services.Add(new ServiceDescriptor(s.ServiceType, s.ImplementationType, ServiceLifetime.Singleton));
		}

		return builder.Build();
	}
}
```

Hope it helped.

**Reply by pjmagee** [reply]
Unfortunately, the Library I was using did not Expose the types. But with a bit of Reflection I managed to unregister and re-register them. Dirty, but it did the job :) Thanks for the response.

**Comment by marwalsch** [comment]
I filed a related [proposal ](https://github.com/dotnet/maui/issues/3622) for this as I kept running into the exact same architectural gap for which the framework unlike Blazor appears to yield to built-in solution.
